Fix: HKTag derivation for unapplied type lambdas with type bounds

AriajSarkar opened the pull request:
/claim #30

This PR fixes an issue where HKTag derivation failed for unapplied type lambdas with type bounds, as reported in #30.

It enables the following usage pattern:
```scala
trait X
trait XAble[_ <: X]
class Y extends X

def getTag[F[_ <: X]: Tag.auto.T] = Tag[F[Y]]

// This now works correctly
getTag[XAble] 
```
Fixes #30.


https://github.com/user-attachments/assets/ac27cfb1-fa2e-42aa-ad4c-0df9ab71673b

another link for 1080p(only for 90 days): https://streamable.com/c569dt

AriajSarkar committed:
Fix #30: TagMacro: make HKTag summonable for unapplied type lambdas with type bounds

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/zio/izumi-reflect?pullRequest=562) <br/>All committers have signed the CLA.

pshirshov commented:
Before I start looking, could you disclose any uses of coding assistants please?
I'm not against them and visually this P/R looks very reasonable. But still, if you used any of them-please tell us.

AriajSarkar commented:
Yeah ChatGPT, it’s mostly for things like the recording title and description.
I’ve never used a screen recorder before and I’m really bad at explaining things, so that’s why.
And yeah, I understand why you asked about the AI, the maintainer of that Rust HTTP crate already told me not to use AI too much because it can get overly verbose sometimes 😅

neko-kai commented:
Update to latest develop for CI fix

AriajSarkar committed:
Merge remote-tracking branch 'upstream/develop' into fix/30-TagMacro

neko-kai commented on izumi-reflect/izumi-reflect/src/test/scala-2/izumi/reflect/test/Issue30Test.scala:
@AriajSarkar Since scala-2 and scala-3 capabilities are now the same wrt Tags with type bounds, you can merge both of the tests of this feature into `SharedTagTest`

AriajSarkar replied:
Got it!

AriajSarkar committed:
Move Issue30 tests into SharedTagTest as requested by maintainer

neko-kai commented on izumi-reflect/izumi-reflect/src/test/scala/izumi/reflect/test/SharedTagTest.scala:
```diff
@@ -214,6 +214,21 @@ abstract class SharedTagTest extends AnyWordSpec with XY[String] with TagAsserti
       assert(b4)
     }
 
+    "support HKTag for unapplied type lambdas with type bounds" in {
```
You may also remove the scala-3 copy of this test case in scala-3 TagTest.scala

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -609,14 +646,13 @@ class TagLambdaMacro(override val c: whitebox.Context) extends TagMacro(c) {
-            silent = false,
-            withImplicitViewsDisabled = true,
-            withMacrosDisabled = true
-          ).tpe
```
this is a meaningless diff, there was no behavior change here

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -465,7 +498,11 @@ class TagMacro(val c: blackbox.Context) {
+    val params = if (tpeParams.size == kind.args.size) {
+      tpeParams.map(p => mkTypeParameter(mutArg, p.typeSignature))
+    } else {
+      kind.args.map(mkTypeParameter(mutArg, _))
```
This is never called because the condition above is always true.
`protected[this] def mkTypeParameter(owner: Symbol, kind: Kind)` is now never used.
~~Note that you could either delete it, or as per my suggestion above,~~ move the TypeBounds info into the Kind type and then delete the other (`def mkTypeParameter(owner: Symbol, tpe: Type): Symbol`) method.

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -453,10 +453,43 @@ class TagMacro(val c: blackbox.Context) {
     tpeSymbol
   }
 
+  protected[this] def mkTypeParameter(owner: Symbol, tpe: Type): Symbol = {
```
`protected[this] def mkTypeParameter(owner: Symbol, kind: Kind)` is now never used. The only condition under which it's called is never true.

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -453,10 +453,43 @@ class TagMacro(val c: blackbox.Context) {
+
+    val tpeTpe = tpe match {
+      case PolyType(params, _) =>
+        val newParams = params.map(p => mkTypeParameter(tpeSymbol, p.typeSignature))
```
I wonder if instead of looping here it wouldn't be more elegant to add TypeBounds to `Kind` data type and extract it in `kindOf`? Then you could in `protected[this] def mkTypeParameter(owner: Symbol, kind: Kind)` merely read the type bounds from `Kind` structure rather than perform pattern match on `Type` inside `mkTypeParameter`

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -453,10 +453,43 @@ class TagMacro(val c: blackbox.Context) {
+      tpe.typeParams
+    } else {
+      tpe.typeSymbol.info match {
+        case PolyType(params, _) => params
```
This condition can only happen when `tpe` is the synthetic type parameter generated in `val ctorParam = mkTypeParameter(NoSymbol, targetTpe)`
This is not obvious from the code. Note that a PolyType always has non-empty typeParams.

~~What you could do instead is pass separate parameters for type constructor (used in `mkPolyType(tpe, params)` below) and for type parameters.~~

```scala
protected[this] def mkHKTagArgStruct(tycon: Type, tpeParams: List[Type], kind: Kind): Type
```

~~Then in `TagLambda` you'd have a call:~~

```scala
val ArgStruct = mkHKTagArgStruct(ctorParam.asType.toType, ctorParam.typeSignature.typeParams, kind)
```

~~And in `summonTagForKind` you'd have a call:~~

```scala
val ArgStruct = mkHKTagArgStruct(tpe, tpe.typeParams, kind)
```

That is, unless, you move type bounds info into `Kind`, then inspecting `tpe.typeParams` will become unnecessary just as before your change and you'd be able to use the same signature as before:

```scala
protected[this] def mkHKTagArgStruct(tycon: Type, kind: Kind): Type
```

AriajSarkar commented:
Thanks for the detailed review!
I’ll remove the Scala-3 duplicate test, clean up the meaningless diff, delete the unused mkTypeParameter overload, and refactor mkHKTagArgStruct following your suggestion (passing explicit tycon and tpeParams / or embedding bounds inside Kind).
I’ll update the PR shortly.

AriajSarkar committed:
fix(TagMacro): revert meaningless diff, update tests

- Restore withImplicitViewsDisabled/withMacrosDisabled flags in lambdaImpl
- Update TagProgressionTest to expect success for Issue #30 fix
- Remove duplicate HKTag test from Scala 3 TagTest.scala

AriajSarkar commented:
`TagProgressionTest` was documenting a previous limitation on Scala 2 that no longer exists after the fix, so I updated it to reflect the new supported behavior.

AriajSarkar committed:
tests: resolve TagProgressionTest merge conflict and update expectations

Merged upstream changes cleanly and preserved new behavior.

AriajSarkar reviewed (commented):
I already fixed the conflict, but GitHub still shows a merge conflict. Can you help me resolve it!?

Edit: Its done!

AriajSarkar committed:
Merge upstream/develop and resolve conflict in TagProgressionTest.scala

neko-kai commented on izumi-reflect/izumi-reflect/src/test/scala-2/izumi/reflect/test/TagProgressionTest.scala:
```diff
@@ -13,18 +11,15 @@ class TagProgressionTest extends SharedTagProgressionTest {
-      val t = intercept[TestFailedException] {
-        assertCompiles("""
-        type `TagK<:Dep`[K[_ <: Dep]] = izumi.reflect.HKTag[ { type Arg[A <: Dep] = K[A] } ]
+    "progression test: type tags with bounds ARE now supported by the macro on Scala 2 (Issue #30 fix)" in {
```
@AriajSarkar Since this is a now a regular test (tests that a feature works), not a progression test (tests that a feature fails to work), it should now be moved to scala-2 TagTest file

AriajSarkar committed:
Move Issue #30 test from TagProgressionTest to TagTest (per maintainer feedback)

neko-kai commented:
Note: I unresolved previous comments about mkTypeParameter, since they weren't resolved. I think the correct solution is to move type bounds info into `Kind` case class and gather it in ReflectionUtil.kindOf - not in-place in mkTypeParameter

AriajSarkar committed:
refactor: move bounds into Kind, update mkTypeParameter & mkHKTagArgStruct as per maintainer review

AriajSarkar commented:
Sorry about that. I’ll apply the full refactor to move type bounds into Kind and update `mkTypeParameter`/`mkHKTagArgStruct `accordingly.

It’s now fully addressed: bounds are moved into Kind, mkTypeParameter uses KindParam, and the older Type-based version has been removed.

All the locations you mentioned have been updated accordingly.

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/ReflectionUtil.scala:
```diff
@@ -139,12 +139,39 @@ private[reflect] object ReflectionUtil {
-  final case class Kind(args: List[Kind]) {
-    def format(typeName: String) = s"$typeName${if (args.nonEmpty) args.mkString("[", ", ", "]") else ""}"
+  /** Stores bounds for a type parameter. lo/hi are None for default bounds (Nothing..Any) */
+  final case class KindParam(
```
There is no point in distinction between KindParam and Kind.

```suggestion
  final case class Kind(
    args: List[Kind],
    bounds: Option[Universe#TypeBounds]
)
```
@AriajSarkar

AriajSarkar replied:
Thanks, that makes sense. Ill drop KindParam and merge the bounds directly into Kind, following your suggested structure. I’ll update the model and the macro helpers accordingly and push the refactor.

AriajSarkar committed:
fix(TagMacro): refactor KindParam usage to simplify type parameter handling

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/ReflectionUtil.scala:
```diff
@@ -139,12 +139,36 @@ private[reflect] object ReflectionUtil {
+        }
+        Kind(kindOf(sig).args, bounds)
+      },
+      None
```
@AriajSarkar A higher-kinded type may also have type bounds. I've added a test case that demonstrates this and works on Scala 3, but fails on current Scala 2 implementation in this branch - https://github.com/zio/izumi-reflect/commit/43746d80273265eb7a041b03bd5977bb999bce71

non-progression form:
```scala
"support HKTag for higher-kinded type lambdas with inner and outer type bounds" in {
  import izumi.reflect.{HKTag, Tag}

  trait Trait3[T <: Dep]
  trait TraitK3[T[x <: Dep] <: Trait3[x]]

  type `TagK<:Dep`[K[_ <: Dep]] = HKTag[{ type Arg[A <: Dep] = K[A] }]
  type `TagKT<:Dep`[K[T[x <: Dep] <: Trait3[x]]] = HKTag[{ type Arg[T[x <: Dep] <: Trait3[x]] = K[T] }]

  def t[K[F[x <: Dep] <: Trait3[x]]: `TagKT<:Dep`, T[x <: Dep] <: Trait3[x]: `TagK<:Dep`] = Tag[K[T]]

  assert(t[TraitK3, Trait3].tag == Tag[TraitK3[Trait3]].tag)
}
```

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -351,7 +351,7 @@ class TagMacro(val c: blackbox.Context) {
         // error: the entire type is just a proper type parameter with no type arguments
         // it cannot be resolved further
-        case Some(Kind(Nil)) =>
+        case Some(k) if k == Kind.`*` =>
```
A proper type with defined type bounds will fail this check, but shouldn't.

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -493,7 +498,7 @@ class TagMacro(val c: blackbox.Context) {
   protected[this] def summonTagForKind(tpe: c.Type, kind: Kind): c.Tree = {
     try {
-      if (kind == Kind(Nil)) {
+      if (kind == Kind.`*`) {
```
A proper type with defined type bounds will fail this check, but shouldn't.

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -631,7 +636,13 @@ class TagLambdaMacro(override val c: whitebox.Context) extends TagMacro(c) {
+      case tb: TypeBoundsApi => Some(tb.asInstanceOf[Universe#TypeBounds])
+      case _ => None
+    }
+    val ctorKind = Kind(kind.args, targetBounds)
```
This is meaningless. `ctorKind` is exactly the same value as `kind` after all these transformations

AriajSarkar commented:
Before I push the update, just wanted to clarify one thing.

The case with both inner and outer bounds (T[x <: Dep] <: Trait3[x]) is basically a separate problem.  
Here the outer bound depends on the inner type parameter, which makes it a lot more complex than the original Issue #30 case.  
The new test you added shows this is a different limitation in Scala 2.

I can’t dive into that part right now because my college exams start on December 10th.  
Totally okay if someone else takes over that piece. Even without the bounty,
I already learned a lot working on this fix.

Also, I ran `sbtgen.sc`. Should I commit the regenerated files, or leave them out of this PR?

neko-kai commented:
@AriajSarkar 
The case with interdependent parameters may indeed be too complicated - it _could_, but I'm not sure whether it does, require reconstructing new TypeBounds with new type parameters from scratch. However there is a less complex case that shouldn't require that but still fails with code in this PR:

```scala
"support HKTag for unapplied type lambdas with higher-kinded type bounds" in {
  class Trait4[K[_]] extends Dep
  class Trait5[T[K[_]]]

  type `TagKT<:Dep`[K[X[_] <: Dep]] = HKTag[{ type Arg[A[_] <: Dep] = K[A] }]

  def t[T[_[_] <: Dep]: `TagKT<:Dep`, A[_] <: Dep: TagK] = Tag[T[A]]

  assert(t[Trait5, Trait4].tag == Tag[Trait5[Trait4]].tag)
}
```

This has bounds in the same place as the original test case, with the only difference that each `_` placeholder became a higher-kinded `_[_]` placeholder. I think this case should be solvable simply by taking bounds from `.finalResultType` when the inspected type has type parameters, but I could be wrong.

- For interdependent bounds case, I noticed that your code seems LLM generated - e.g. making copies of methods and types instead of refactoring is a specific trait of Claude, not to mention the excessive comments - you could try running a few more prompts to try and fix the interdependent case. If you're using latest Opus 4.5 or Gemini 3, I believe these LLMs have advanced enough to find a solution. If it doesn't find a solution - OK, non-recursive cases are enough in practice, that case would then remain as a progression test.

- For sbtgen.sc, committed build.sbt should be generated with `./sbtgen.sc --js --native` (so that Scala Steward can find and update Native and JS dependencies) – the diff should evaporate if you regenerate it like that. You can use a JVM-only config for development, but the committed version should be --js --native. There are no changes to dependencies in this PR and none are required, so there is no good reason to add regenerated build files to the PR.

- Re: others taking your bounty - not unless you disappear for 3+ months. This is fairly sleepy project, no rush.

AriajSarkar commented:
Thanks for the clarification -- that actually helps a lot.

I’ll take a look at the higher-kinded version of the original case (the K[X[_] <: Dep] one).  
It feels close enough to the first fix, so I should be able to tweak the bounds extraction and use finalResultType where it makes sense.

The interdependent-bounds case is definitely a bigger jump.  
After my exams I can try exploring it a bit, but I can’t promise anything solid there right now. 😅

And noted about sbtgen.sc - I’ll regenerate with `--js --native` locally but won’t commit build.sbt since this PR doesn’t touch any dependencies.

Also, thanks for the reassurance about the bounty timeline. 🙂

AriajSarkar committed:
fix(TagMacro): enhance HKTag support for higher-kinded type lambdas with type bounds

AriajSarkar commented:
Update on the more complicated case with the mixed inner/outer bounds.

I tweaked the fix again and Scala 2 is now handling these two situations properly:

* simple bounds like `F[_ <: Dep]`
* and the HK case `K[X[_] <: Dep]`

What I changed basically boils down to:
`kindOf` now pulls the bounds either from `TypeBounds` or from the `PolyType.resultType`, and I stash those inside the internal `Kind` thing. Then `mkTypeParameter` re-applies the bounds when it builds the Arg structure. With that, HKTag derivation behaves the way it’s supposed to.

After that I tried to push it further into the “fully interdependent” situation, something like this:

```scala
"support HKTag for higher-kinded type lambdas with inner and outer type bounds" in {
  import izumi.reflect.{HKTag, Tag}

  trait Trait3[T <: Dep]
  trait TraitK3[T[x <: Dep] <: Trait3[x]]

  type `TagK<:Dep`[K[_ <: Dep]] =
    HKTag[{ type Arg[A <: Dep] = K[A] }]

  type `TagKT<:Dep`[K[T[x <: Dep] <: Trait3[x]]] =
    HKTag[{ type Arg[T[x <: Dep] <: Trait3[x]] = K[T] }]

  def t[K[F[x <: Dep] <: Trait3[x]]: `TagKT<:Dep`,
        T[x <: Dep] <: Trait3[x]: `TagK<:Dep`] =
    Tag[K[T]]

  assert(t[TraitK3, Trait3].tag == Tag[TraitK3[Trait3]].tag)
}
```

The issue becomes obvious when you stare at it: the same type parameter (`x`) is used in the bound (`x <: Dep`) **and** inside `Trait3[x]`.
But the HKTag encoding introduces a *new* binder when it does `{ type Arg[...] = K[A] }`, and there’s no way to force that “A” to actually be the *same* `x` used in the inner structure. So the type system can’t represent that interdependent relationship cleanly. It's basically a limitation of the encoding itself, not a missing piece in the implementation.

So right now the situation is:

* simple bounds -> works
* HK bounds like `K[X[_] <: Dep]` -> works
* genuinely interdependent inner + outer bounds → still not representable with HKTag (Scala just doesn’t give us enough expressive power here)

Everything added to `kindOf` and `mkTypeParameter` is solid for the cases Scala 2 can actually express, and all 207 tests pass.

neko-kai commented:
@AriajSarkar 
> there’s no way to force that “A” to actually be the same x used in the inner structure. So the type system can’t represent that interdependent relationship cleanly

Just tell the LLM that it's confused by syntax - the x on the RHS is not, cannot be and does not have to be the same as on the LHS. The assertion that the type system can't represent the relationship is disproven a priori by the fact that it typechecks. The assertion that it's generally inexpressible is disproven by the fact that Scala 3 equivalent works right now.

```scala
    "Scala 3 supports Tag for higher-kinded type lambdas with inner and outer type bounds" in {
      import izumi.reflect.Tag

      trait Trait3[T <: Dep]
      trait TraitK3[T[x <: Dep] <: Trait3[x]]

      type `TagK<:Dep`[K[_ <: Dep]] = Tag[K]

      type `TagKT<:Dep`[K[T[x <: Dep] <: Trait3[x]]] = Tag[K]

      def t[K[F[x <: Dep] <: Trait3[x]]: `TagKT<:Dep`, T[x <: Dep] <: Trait3[x]: `TagK<:Dep`] = Tag[K[T]]

      assert(t[TraitK3, Trait3].tag == Tag[TraitK3[Trait3]].tag)
    }
```

Tell it to keep working in a loop until it can write a working solution or until it can form a consistent theory as to why it can't do that, one that's not immediately disproven by a working Scala 3 example.

AriajSarkar committed:
refactor(TagTest): add new HKTag test for higher-kinded type lambdas with interdependent bounds

AriajSarkar commented:
I pushed the updated changes.
Can you check if there are any remaining issues or anything I still got wrong?  
If something needs adjusting, just tell me and I’ll fix it.

neko-kai committed:
Merge branch 'develop' into fix/30-TagMacro

neko-kai commented:
@AriajSarkar 
Wow, looks like you actually did it! 👍 

There is now however a StackoverflowError if you attempt to reformulate the test using `Tag.auto.T`:

```scala
"support combining with Tag for higher-kinded type lambdas with inner and outer type bounds" in {
  trait Trait3[T <: Dep]
  trait TraitK3[T[x <: Dep] <: Trait3[x]]

  def t[K[F[x <: Dep] <: Trait3[x]]: Tag.auto.T, T[x <: Dep] <: Trait3[x]: Tag.auto.T] = Tag[K[T]]

  assert(t[TraitK3, Trait3].tag == Tag[TraitK3[Trait3]].tag)
}
```

A copy of this test is currently in scala-3 TagTest.scala : fixing StackoverflowError will allow to move this test to `SharedTagTest`, because then both Scala 3 and 2 will be able to write this test with exactly the same syntax.

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/ReflectionUtil.scala:
```diff
@@ -139,12 +139,41 @@ private[reflect] object ReflectionUtil {
+    * @param args nested kinds for each type parameter
+    * @param bounds optional type bounds (None for default bounds Nothing..Any)
+    */
+  final case class Kind(args: List[Kind], bounds: Option[Universe#TypeBounds]) {
```
The CI is failing due to MiMa rules - to fix it, add a rule to Ignore changes in `ReflectionUtil` object to `project/Deps.sc` and regenerate build - ReflectionUtil is already package private and shouldn't be checked anyway. There are plenty of examples, look for `ProblemFilters.exclude[Problem]` in Deps.sc

AriajSarkar committed:
fix(build.sbt): exclude ReflectionUtil from Mima problem filters

AriajSarkar committed:
fix(TagMacro): handle identity-like applications in HKTag resolution

AriajSarkar committed:
fix(TagMacro): improve handling of identity-like applications and enhance HKTag support for higher-kinded type lambdas

AriajSarkar committed:
Add MiMa filter for ReflectionUtil (package-private)

AriajSarkar commented:
@neko-kai Done!

neko-kai committed:
revert 'Identity application' changes from e508200f7edc3312a369f5e597bbfccfd6acd640 ; add comments to clarify what happens now ; remove makeHKTagFromTpe method

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -134,25 +134,33 @@ class TagMacro(val c: blackbox.Context) {
+        // outerLambda is a simple forwarder `[A, B, C] =>> F[A, B, C]` AND `F` is a type parameter
+        // We can't progress here: we must find HKTag[F] to progress, but we ARE in the process of
+        // deriving HKTag[F], which means it doesn't already exist.
+        case Some(_) if outerLambdaEqualsCtorEtaExpand =>
```
@AriajSarkar I reverted the changes you made in https://github.com/AriajSarkar/izumi-reflect/commit/e508200f7edc3312a369f5e597bbfccfd6acd640 - `return makeHKTagFromStrongTpe[ArgStruct](outerLambda)` caused a Tag with unresolved type parameters to be created, with reference to unresolved type parameter `izumi.reflect.test.SharedTagProgressionTest.F`. We can't create these tags. Moreover, we can't create a Tag when `outerLambdaEqualsCtorEtaExpand` is true, I explained why in the code comments – because when we encounter a construction where `outerLambda` = `[A] =>> F[A]` and `ctorTpe` is `F` and `F` is a type parameter, not a concrete type, that means we're in the process of trying to materialize a `Tag` for F. But we can't materialize a Tag for a type parameter, it has to be passed at runtime by user (`F[_]: TagK`), if we ended up in this branch inside the macro, we can only fail.

AriajSarkar replied:
Yeah, I get it now 🫠
My change was basically trying to create a Tag while the type parameter was still unresolved ,, that test’s F wasn’t actually bound yet, so the macro ended up building a Tag for something that doesn’t exist in that spot.

And in the outerLambdaEqualsC.... case, we’re literally inside the point where HKTag[F] is being derived, so there’s no way to “progress” there without breaking things. Your original comment made more sense after looking at it again 😅.

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/TagMacro.scala:
```diff
@@ -434,27 +437,35 @@ class TagMacro(val c: blackbox.Context) {
 
-    setInfo(tpeSymbol, tpeTpe)
+    val tpe = if (kind.args.nonEmpty) {
+      val params = kind.args.zipWithIndex.map { case (k, i) => mkTypeParameter(sym, k, origInner.lift(i)) }
```
@AriajSarkar Could you refactor to get rid of this zip? Just add pt.typeParams to Kind structure too, together with bounds. That way everything you'll need in mkTypeParameter will be in Kind and you wouldn't have to reinspect the type again inside mkTypeParameter and zip new data back together with what was in Kind - which makes for ugly code.

AriajSarkar replied:
Ok👍

neko-kai commented:
@AriajSarkar I think all you have to do now is refactor the code to make it more readable before merging, as per my review comment, and that'll be it 👍

AriajSarkar committed:
Add typeParams to Kind structure to eliminate zipWithIndex pattern

AriajSarkar commented:
@neko-kai Done!

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/ReflectionUtil.scala:
```diff
@@ -139,12 +139,47 @@ private[reflect] object ReflectionUtil {
+    * @param bounds optional type bounds (None for default bounds Nothing..Any)
+    * @param typeParams original type parameter symbols (for substitution during type construction)
+    */
+  final case class Kind(args: List[Kind], bounds: Option[Universe#TypeBounds], typeParams: List[Universe#Symbol] = Nil) {
```
@AriajSarkar The last refactoring made this type incoherent. it doesn't make sense to have a separate field for typeParams, when args already represents typeParams. This should work:
```suggestion
  final case class Kind(args: List[Kind], bounds: Option[Universe#TypeBounds], symbol: Universe#Symbol) {
```

AriajSarkar replied:
Makes sense. I see how args already covers that, so separating typeParams is redundant.

AriajSarkar replied:
wait i need to check locally.

AriajSarkar committed:
Update izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/ReflectionUtil.scala

Co-authored-by: Kai <450507+neko-kai@users.noreply.github.com>

AriajSarkar committed:
refactor: change Kind.typeParams to Kind.symbol per review

- Store symbol per Kind node instead of list at root level

- Update kindOf to set each param's symbol on its inner Kind

- Use NoSymbol sentinel for root Kind and companion methods

- Update TagMacro.mkTypeParameter to access symbols via kind.args

AriajSarkar commented:
@neko-kai Done! Sir.

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/ReflectionUtil.scala:
```diff
@@ -139,12 +139,45 @@ private[reflect] object ReflectionUtil {
   }
 
+  object Kind {
+    private val noSymbol: Universe#Symbol = scala.reflect.runtime.universe.NoSymbol
```
Runtime universe does not match the universe used by macro expansion, if this noSymbol is actually used, it may blow up at runtime

neko-kai commented on izumi-reflect/izumi-reflect/src/main/scala-2/izumi/reflect/ReflectionUtil.scala:
```diff
@@ -139,12 +139,45 @@ private[reflect] object ReflectionUtil {
+        Kind(kindOf(sig).args, bounds, param.asInstanceOf[Universe#Symbol])
+      },
+      None,
+      scala.reflect.runtime.universe.NoSymbol
```
@AriajSarkar The outermost 'Kind' type constructor has both bounds and a type symbol. They may be extracted reusing the exact same code you use for parameters.
Given
`def kindOf(tpeOrSym: Either[Universe#Type, Universe#Symbol]): Kind = {`

Then 
`val sig = tpeOrSym.fold(identity, _.typeSignature`
and
`val tpe`/`val param = tpeOrSym.fold(_.typeSymbol, identity)`

neko-kai commented:
I fixed the remaining code smells in https://github.com/zio/izumi-reflect/commit/636ab10072cf9edbfa0be35627962b6e59a722ec and https://github.com/zio/izumi-reflect/commit/6c65683e94d33a02a3df303eca9cc7075543b734

AriajSarkar commented:
Oh thanks sir, I didn’t see the message because I was working on a separate project. Sorry for not replying. I’ll read the codes tomorrow since it’s 3 AM 🫠 in india, I need to sleep.

pshirshov commented:
@AriajSarkar could I suggest you to try this issue: https://github.com/zio/izumi-reflect/issues/350 - it should be much simpler than this one?

AriajSarkar commented:
@pshirshov Sorry for the late reply, Sir! My sem exams started on Dec 10th 😅.

I did manage to look into the issue after my exam today, especially this explanation: https://github.com/zio/izumi-reflect/pull/553#issuecomment-3224231605

I think I have the core idea down for #350: 3 levels of caching (TagMacro, LTT, database), controlled by flags, to help us benchmark the best default.

I’ll need a little time to write down the plan properly. My exams end on December 22nd 🥲, but I'll try to prep some notes in the meantime so I can push a basic version for review as soon as I'm done.
